### PR TITLE
Only output interior cells

### PR DESCRIPTION
**Pull request by totto82**
This fixes parallel output of FLOWS and FLORES

It should also give a small speed up of the output part in parallel.

### CODE REVIEW

**Comment by totto82** [comment]
jenkins build this please

File: ebos/eclgenericoutputblackoilmodule.cc [hunk 1]
```diff
@@ -715,8 +715,13 @@ regionSum(const ScalarBuffer& property,
         if (property.empty())
             return totals;
 
-        assert(regionId.size() == property.size());
-        for (std::size_t j = 0; j < regionId.size(); ++j) {
+        // the regionId contains the ghost cells
+        // the property does not contain the ghostcells
+        // This code assumes that that the ghostcells are
+        // added after the interior cells
+        // Is this a valid assumption?
```
Comment (bska) [inline ebos/eclgenericoutputblackoilmodule.cc:722]: The assumption (interior cells stored before ghost cells) holds unless the user changes the `OwnerCellsFirst` runtime parameter from its default state of `true`.

**Review by bska — COMMENTED** [review]
It looks like this does the right thing in the [default](https://github.com/OPM/opm-simulators/pull/2512) case of `OwnerCellsFirst = true`.  Whether or not we want to continue to support `OwnerCellsFirst = false` is maybe a matter of debate.

File: ebos/eclgenericoutputblackoilmodule.cc (hunk 1 above)
Comment (atgeirr) [reply ebos/eclgenericoutputblackoilmodule.cc:722]: I think testing and throwing if not the case may be in order then? Or even removing the OwnerCellsFirst runtime option.

Comment (blattms) [reply ebos/eclgenericoutputblackoilmodule.cc:722]: Just remove the option.

Comment (blattms) [reply ebos/eclgenericoutputblackoilmodule.cc:722]: I correct my statement: Maybe someone should first test what happens with grids apart from CpGrid.

Comment (totto82) [reply ebos/eclgenericoutputblackoilmodule.cc:722]: The linear solver has
```
            const bool ownersFirst = EWOMS_GET_PARAM(TypeTag, bool, OwnerCellsFirst);
            if (!ownersFirst) {
                const std::string msg = "The linear solver no longer supports --owner-cells-first=false.";
                if (on_io_rank) {
                    OpmLog::error(msg);
                }
                OPM_THROW_NOLOG(std::runtime_error, msg);
            }
```
I can add something similar to the output code.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
Can we merge this now?

File: ebos/eclwriter.hh [hunk 2]
```diff
@@ -329,15 +331,16 @@ public:
         }
 
         const auto& gridView = simulator_.vanguard().gridView();
-        const int numElements = gridView.size(/*codim=*/0);
+        const int num_interior = Opm::detail::
+            countLocalInteriorCellsGridView(gridView);
```
Comment (bska) [inline ebos/eclwriter.hh:335]: We should generally not use the `Opm::` namespace qualifier when we're inside that namespace.  We should either just drop that qualifier (i.e., `detail::countLocalInteriorCellsGridView()`, preferable) or use `::Opm::` instead (`::Opm::detail::count*()`).

File: ebos/ecloutputblackoilmodule.hh [hunk 3]
```diff
@@ -173,6 +173,16 @@ public:
...
+
+        const bool ownersFirst = EWOMS_GET_PARAM(TypeTag, bool, OwnerCellsFirst);
+        const bool on_io_rank = (simulator.gridView().comm().rank() == 0);
+        if (!ownersFirst) {
+            const std::string msg = "The output code does not support --owner-cells-first=false.";
+            if (on_io_rank) {
+                OpmLog::error(msg);
+            }
+            OPM_THROW_NOLOG(std::runtime_error, msg);
+        }
```
Comment (bska) [inline ebos/ecloutputblackoilmodule.hh:185]: I think I'd probably write this as
```C++
if (! EWOMS_GET_PARAM(TypeTag, bool, OwnerCellsFirst)) {
    const std::string msg = ...;
    if (collectToIORank.isIORank()) {
        OpmLog::error(msg);
    }
    OPM_THROW_NOLOG(std::runtime_error, msg);
}
```
instead.

File: ebos/damariswriter.hh [hunk 4]
```diff
@@ -323,17 +323,18 @@ private:
         }
 
         const auto& gridView = simulator_.vanguard().gridView();
-        const int numElements = gridView.size(/*codim=*/0);
+        const int num_interior = Opm::detail::
+            countLocalInteriorCellsGridView(gridView);
```
Comment (bska) [inline ebos/damariswriter.hh:327]: Please avoid using the `Opm::` namespace qualifier when inside that namespace.

**Review by bska — COMMENTED** [review]
> Can we merge this now?

Just a couple of minor nits remaining, I think.

**Comment by totto82** [comment]
jenkins build this please

**Comment by totto82** [comment]
jenkins build this please

**Review by bska — APPROVED** [review]
Thanks a lot for the updates.  This looks good to me now and I'll merge into master.  At some point we should probably just remove the `OwnerCellsFirst`  option.

**Pull request merged**